### PR TITLE
Don't require pegasus developers to precompile dashboard assets

### PR DESCRIPTION
**Pull request by davidsbailey**
Teacher dashboard was changed to rely on digested and possibly minified dashboard assets in https://github.com/code-dot-org/code-dot-org/pull/13340. This broke developers who hadn't run `rake assets:precompile` recently. This PR eliminates the need for developers to precompile assets by falling back to uncompiled assets when an asset map is unavailable or does not contain a certain asset.

I've verified locally that after removing the contents of `dashboard/public/assets`, teacher dashboard will fall back to serving js from `apps/build/package`. This still requires `npm run build` to be run before loading teacher dashboard, which is unavoidable now that teacher dashboard relies on apps js.

### CODE REVIEW

File: pegasus/test/test_asset_helpers.rb [hunk 1]
```diff
@@ -2,26 +2,100 @@
 require_relative '../../lib/cdo/pegasus'
 require 'minitest/autorun'
 
-CDO.dashboard_assets_dir = './test/fixtures/dashboard_assets'
+# AssetMap must be re-initialized before each test with CDO.dashboard_assets_dir set properly
```
Comment (aoby) [inline pegasus/test/test_asset_helpers.rb]: Is this comment only relevant to development mode? Please clarify. I'm not sure what the action here is, nor what "set properly" means.

File: pegasus/test/test_asset_helpers.rb [hunk 2]
```diff
@@ -2,26 +2,100 @@
...
+      assert_equal MINIFIED_ASSET_PATH, minifiable_asset_path(UNMINIFIED_ASSET_NAME), "incorrect minifiable asset path"
+    end
+  end
+
+  def test_production_no_asset_map
+    # This directory does not have a manifest file in it
+    CDO.stub(:dashboard_assets_dir, './test/fixtures/') do
+      Singleton.__init__(AssetMap)
+      CDO.stub(:rack_env, :production) do
+        CDO.stub(:pretty_js, false) do
```
Comment (aoby) [inline pegasus/test/test_asset_helpers.rb]: Note you can also use [Mocha](http://gofreerange.com/mocha/docs/) [stubs](http://gofreerange.com/mocha/docs/Mocha/Mock.html#stubs-instance_method). They are scoped to the current test case, so will be equivalent to these minitest [stub](http://www.rubydoc.info/gems/minitest/4.2.0/Object:stub) calls, without need for all the nested blocks. 

e.g.
```ruby
CDO.stubs(dashboard_assets_dir: './test/fixtures/')
#...
```

File: pegasus/test/test_asset_helpers.rb (hunk 1 above)
Comment (davidsbailey) [reply pegasus/test/test_asset_helpers.rb]: I've removed this comment. I was trying to explain that the singleton needs to be initialized before each test run, however there's no need for this warning since it will be impossible for someone to ignore this without breaking the tests.

File: pegasus/test/test_asset_helpers.rb [hunk 3]
```diff
@@ -1,27 +1,99 @@
...
+          begin
+            asset_path(UNMINIFIED_ASSET_NAME)
+            raise 'Expected asset_path to raise when asset map not initialized'
+          rescue Exception => e
+            assert_equal e.message, 'Asset map not initialized'
+          end
```
Comment (aoby) [inline pegasus/test/test_asset_helpers.rb]: I don't get this. Why raise the exception in the test and rescue it immediately?

File: pegasus/test/test_asset_helpers.rb [hunk 4]
```diff
@@ -1,27 +1,99 @@
...
+      end
+    end
+  end
+
+  def test_production_asset_not_in_map
+    CDO.stub(:rack_env, :production) do
+      CDO.stub(:pretty_js, false) do
+        begin
+          asset_path(UNMINIFIED_ASSET_NOT_IN_MAP)
+          raise 'Expected asset_path to raise when asset not found'
```
Comment (ashercodeorg) [inline pegasus/test/test_asset_helpers.rb]: This could be much more clearly written as `assert_raise ...`. Also below.

File: pegasus/test/test_asset_helpers.rb (hunk 3 above)
Comment (aoby) [reply pegasus/test/test_asset_helpers.rb]: Also it's better form to rescue the more specific `RuntimeError`, or even `StandardError` instead of `Exception`. RuntimeError is the default exception type with [raise](http://apidock.com/ruby/Kernel/raise), and the base Exception includes system-level stuff we probably don't want to rescue. See http://rubylearning.com/images/exception.jpg

Comment (aoby) [reply pegasus/test/test_asset_helpers.rb]: And for tests, follow this pattern:

```ruby
e = assert_raises RuntimeError do
  # code that is expected to raise an error
end
assert_equal 'my expected message', e.message

File: pegasus/test/test_asset_helpers.rb (hunk 2 above)
Comment (davidsbailey) [reply pegasus/test/test_asset_helpers.rb]: ah yes, `stubs` is much better!

**Comment by davidsbailey** [comment]
Thanks for the detailed review @aoby . Please take another look.

File: pegasus/test/test_asset_helpers.rb [hunk 5]
```diff
@@ -1,27 +1,81 @@
...
   def test_asset_map_pretty
-    CDO.stub(:pretty_js, true) do
-      assert_equal UNMINIFIED_ASSET_PATH, asset_path(UNMINIFIED_ASSET_NAME), "incorrect unminified asset path"
-      # Should return the unminified path because pretty_js is true
-      assert_equal UNMINIFIED_ASSET_PATH, minifiable_asset_path(UNMINIFIED_ASSET_NAME), "incorrect minified asset path"
-    end
+    CDO.stubs(:pretty_js).returns(true)
+    assert_equal UNMINIFIED_ASSET_PATH, asset_path(UNMINIFIED_ASSET_NAME), "incorrect unminifiable asset path"
+    # Should return the unminified path because pretty_js is true
+    assert_equal UNMINIFIED_ASSET_PATH, minifiable_asset_path(UNMINIFIED_ASSET_NAME), "incorrect minifiable asset path"
```
Comment (ashercodeorg) [inline pegasus/test/test_asset_helpers.rb]: super nit: Line length of 80 (throughout).
```
assert_equal(
   UNMINIFIED_ASSET_PATH,
  minifiable_asset_path(UNMINIFIED_ASSET_NAME),
  "incorrect minifiable asset path"
)
```
not a nit: Aside, I don't see what is gained from the custom error message. A failure should tell me that.

File: pegasus/test/test_asset_helpers.rb [hunk 6]
```diff
@@ -1,27 +1,81 @@
...
-      assert_equal MINIFIED_ASSET_PATH, minifiable_asset_path(UNMINIFIED_ASSET_NAME), "incorrect minified asset path"
+    CDO.stubs(:pretty_js).returns(false)
+    assert_equal UNMINIFIED_ASSET_PATH, asset_path(UNMINIFIED_ASSET_NAME), "incorrect unminifiable asset path"
+    assert_equal MINIFIED_ASSET_PATH, minifiable_asset_path(UNMINIFIED_ASSET_NAME), "incorrect minifiable asset path"
+  end
+
+  def test_production_no_asset_map
+    # This directory does not have a manifest file in it
+    CDO.stubs(:dashboard_assets_dir).returns('./test/fixtures/')
+    Singleton.__init__(AssetMap)
```
Comment (ashercodeorg) [inline pegasus/test/test_asset_helpers.rb]: This is removable (being part of `setup`)? Also below.

File: pegasus/test/test_asset_helpers.rb (hunk 5 above)
Comment (davidsbailey) [reply pegasus/test/test_asset_helpers.rb]: I see what you're saying. Custom error messages are a habit for me to reliably indicate which assert is failing, preventing off-by-1 errors in determining which line number an assert is coming from.

File: pegasus/test/test_asset_helpers.rb (hunk 6 above)
Comment (davidsbailey) [reply pegasus/test/test_asset_helpers.rb]: Not removable, because its setting a different value from the one in `setup`. The comment on the previous line alludes to this. Unfortunately I can't give a non-existent directory or it will fail.

Comment (ashercodeorg) [reply pegasus/test/test_asset_helpers.rb]: Ah right. Perhaps (for the slow like me):
`# This directory does not have a manifest file in it. Update the AssetMap.`

Comment (ashercodeorg) [reply pegasus/test/test_asset_helpers.rb]: Or not, LGTM.
